### ISSUE TITLE
Expose API to signify if an `or-expression` (`|`) surprisingly widened a sign extended value.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#56994]
See https://github.com/dotnet/roslyn/issues/56977 for motivating scenario.  Currently the compiler warns for code like:

```c#
        var random = new Random();
        long result = random.Next();
        result <<= 32;
        result |= random.Next(); // warning here
        return result;
```

this impacts the IDE's ability to determien if we should suggest removing the following cast:

```c#
        var random = new Random();
        long result = random.Next();
        result <<= 32;
        result |= (long)random.Next();
        return result;
```

Semantics and runtime behavior are identical (so we think we should) but the removal causes this warning.  Replicating the compiler logic for this results in 300+ lines of duplicated compiler code plopped into the IDE.

Proposed api for this is implemented here and has been validated to work: https://github.com/dotnet/roslyn/pull/56992

The proposed API is:

```c#
    public static class IBinaryOperationExtensions
    {
        /// <summary>
        /// Returns true if this is a bitwise-or (<c>|</c>) of two values, at least one of which is signed and 
        /// will be widened in the process.  For example <c>someLong | someInt</c>.  This sort of operation
        /// can lead to surprising results when negative values are involved, as sign extension of a negative integer
        /// value fills the extended bits all with <c>1</c>, and thus will overwrite all the high order bits as
        /// well in the result.  While legal, it is recommended that code explicitly cast the smaller value to
        /// clearly indicate the extending approach desired (either sign extended or not).
        /// </summary>
        public static bool IsBitwiseOrOfSignExtendedOperand(this IBinaryOperation binaryOperation, CancellationToken cancellationToken = default);
    }

    public static class ICompoundAssignmentOperationExtensions
    {
        /// <summary>
        /// Returns true if this is a bitwise-or-assignment (<c>|=</c>) of two values, at least one of which is signed and 
        /// will be widened in the process.  For example <c>someLong |= someInt</c>.  This sort of operation
        /// can lead to surprising results when negative values are involved, as sign extension of a negative integer
        /// value fills the extended bits all with <c>1</c>, and thus will overwrite all the high order bits as
        /// well in the result.  While legal, it is recommended that code explicitly cast the smaller value to
        /// clearly indicate the extending approach desired (either sign extended or not).
        /// </summary>
        public static bool IsBitwiseOrOfSignExtendedOperand(this ICompoundAssignmentOperation assignmentOperation, CancellationToken cancellationToken = default);
    }
```

Open design questions are:

1. What should the name of this be? 
2. Should there be two methods (strongly typed), or one method on IOperation for this?
  a. if htere is one operation should it throw if not passed a BinOp/CompoundOp, or should it just return false?
